packages/cardano-services/src/TxWorker/utils.ts

```diff
@@ -0,0 +1,100 @@
+/* eslint-disable @typescript-eslint/no-explicit-any */
+/* eslint-disable @typescript-eslint/no-shadow */
+/* eslint-disable no-use-before-define */
+import { CONNECTION_ERROR_EVENT, TxSubmitWorker } from '@cardano-sdk/rabbitmq';
+import { CommonOptionDescriptions } from '../ProgramsCommon';
+import { DnsResolver, MissingProgramOption, ServiceNames, srvRecordToRabbitmqURL } from '../Program';
+import { Logger } from 'ts-log';
+import { TxSubmitProvider } from '@cardano-sdk/core';
+import { TxWorkerOptions } from './loadTxWorker';
+
+type WorkerFactory = () => Promise<TxSubmitWorker>;
+
+/**
+ * Create a worker factory with service discovery
+ *
+ * @param {DnsResolver} dnsResolver used for dns resoluton
+ * @param {TxSubmitProvider} txSubmitProvider tx submit provider 'ogmiosTxSubmitProvider'
+ * @param {Logger} logger common logger
+ * @param {TxWorkerOptions} options needed for tx worker initialization
+ * @returns {WorkerFactory} WorkerFactory with service discovery, returning a 'TxSubmitWorker' instance
+ */
+export const createWorkerFactoryWithDiscovery =
+  (
+    dnsResolver: DnsResolver,
+    txSubmitProvider: TxSubmitProvider,
+    logger: Logger,
+    options: TxWorkerOptions
+  ): WorkerFactory =>
+  async () => {
+    const record = await dnsResolver(options.rabbitmqSrvServiceName!);
+    return new TxSubmitWorker(
+      { ...options, rabbitmqUrl: srvRecordToRabbitmqURL(record) },
+      { logger, txSubmitProvider }
+    );
+  };
+
+/**
+ * Create and start a new worker instance with registered listener for connection error events
+ *
+ * @param {WorkerFactory} workerFactory creates a worker instance with service discovery
+ * @returns {TxSubmitWorker} TxSubmitWorker instance
+ */
+export const createAndStartNewWorker = async (workerFactory: WorkerFactory) => {
+  const newWorker = await workerFactory();
+  attachOnConnectionErrorHandler(newWorker, workerFactory);
+  await newWorker.start();
```

Comment:
This can throw. We might want to apply exponential backoff retry strategy to re-resolve and recreate the worker when it happens

Reply:
Depends on error. If it's a connection error then I think we want to wait for a second and re-resolve.

Reply:
in addition: if `start()` throw caused by connection error - the worker will emit a new event and the abstraction will try to re-resolve and [create new instance](https://github.com/input-output-hk/cardano-js-sdk/blob/89a1a0d3cf617df8746f6493e70edfa219226835/packages/cardano-services/src/TxWorker/utils.ts#L73)
Could you check our discussion [here](https://github.com/input-output-hk/cardano-js-sdk/pull/344#discussion_r926754387) and say your opinion about the emitting events for all kind of errors, I think it makes sense

Reply:
It will emit the event, but it will also throw an error which I believe is an unhandled promise rejection right now.

> and say your opinion about the emitting events for all kind of errors instead of only connection-related ones

I think it should just throw (to terminate the process) on any other errors and rely on external infrastructure to restart. If there's some unexpected error state (e.g. faulty/destroyed txSubmitProvider) re-resolving might not help. Running forever in a faulty state would be pretty bad.

Reply:
Based on our brainstorming at today's daily, the error handling refactoring of `TxSubmitWorker` is introduced within https://github.com/input-output-hk/cardano-js-sdk/pull/344/commits/5d02656de2f4649f9528084a9dfcbe2b49990a7d , please review the changes when you have time @mkazlauskas  @iccicci  :)